Channel Assignment TrackBuilder output Tracks + Stubs and InputRouter input Stubs

tschuh opened the pull request:
This PR does:
- updates TrackBuilderChannel::channelId using now TTTrackRef instead of TTTrack
- extends TrackBuilderChannel to assign TTStubs from found TTTracks to channel
- trackFindingTracklet::ProducerKFin uses now TTTrackRefs to get channelIds via TrackBuilderChannel
- ProducerTrackBuilderCannel construct TrackBuilderChannel with Setup which is needed to assign stubs to channel
- adds a map of which DTC links are used in summer chain configuration as python config
- adds seed type names, seeding layer Ids per seed type and projection layers per seed type as python config
I add constants which are somehow somewhere already present in the tracklet pattern reco c++ to python configuration. This needs rethinking, advise is welcome.

tschuh committed:
track builder channel upgraded to assign stubs to channel.

tschuh committed:
code format

tschuh committed:
readded missing pieces.

tomalin commented on L1Trigger/TrackFindingTracklet/python/ProducerKF_cfi.py:
```diff
@@ -13,6 +14,8 @@
-  EnableTruncation     = cms.bool    ( True  )                                                   # enable emulation of truncation, lost stubs are filled in BranchLost
+  EnableTruncation     = cms.bool    ( True  ),                                                  # enable emulation of truncation, lost stubs are filled in BranchLost
 
+  SummerChain         = cms.bool  ( False ),                                                    # reduce l1 tracking to summer chain configuration
```
The summer chain will only exist for a few more months. So we should minimize and localise as much as possible any code that is specific to the summer chain. And try to generalize it so it will work with larger chains, which we'll soon be testing.
Also, it's messy if we need to set SummerChain=true here and in Settings.h
Also, do SummerChainChannels represent which input channels of Apollo are connected to DTC? If so, why is this in the KF config file? Where do these numbers come from?

tschuh replied:
I can't easily retrieve the Reduced bool from Settings in other ED or ES produces as L1FPGATrackProducer. Using an ESProducer to produce and manipulate Settings would be very handy. As long as that is as it is we will have to live with duplicated settings.
I can replace the name SummerChain with Reduced if that makes you happy.
I can create a new cfi file for the vector of dtc channel if that makes you happy as well.
I translated them by hand.

tomalin commented on L1Trigger/TrackFindingTracklet/python/ProducerTrackBuilderChannel_cfi.py:
```diff
@@ -2,9 +2,31 @@
+    PtBoundaries        = cms.vdouble( 1.34 ),                                                 # pt Boundaries in GeV, last boundary is infinity
+
+    MaxNumProjectionLayers    = cms.int32( 8 ), # max number layers a sedd type may project to
+    SeedTypesSeedLayers       = cms.PSet (      # seeding layers of seed types using default layer id [barrel: 1-6, discs: 11-15]
```
The seeding & projection layers are already defined in Settings.h (albeit in a slightly different format to in this python file). So we could take these constants from there, to avoid specifying them in two different places. 
If we do choose to keep them in python, since they are needed by many parts of tracking code, they should be moved to TrackTrigger/python/ProducerSetup_cfi.py .
From the discussion in https://hypernews.cern.ch/HyperNews/CMS/get/sw-develtools/2701.html , I conclude that where params are shared with the HLS code, such as here, their default values should be specified in Settings.h rather than in python cfg. I therefore recommend removing these from python cfg.

tschuh replied:
done by rename those as subsets

tschuh committed:
renamed TrackBuilderChannel to ChannelAssignment now also containing DTC to IR mapping

tschuh commented:
Renamed TrackBuilderChannel to ChannelAssignment. One can now run any reduced chains by manipulating ChannelAssignment_cfi.py SeedTypes, MaxNumProjectionLayers, SeedTypesSeedLayers, SeedTypesProjectionLayers and IRChannelsIn.

tschuh committed:
name change in L1Trigger/TrackFindingTracklet/python/ProducerKF_cff.py

tschuh committed:
bug fix

tomalin commented on L1Trigger/TrackFindingTracklet/interface/ChannelAssignment.h:
```diff
@@ -0,0 +1,63 @@
+namespace trackFindingTracklet {
+
+  /*! \class  trackFindingTracklet::ChannelAssignment
+   *  \brief  Class to assign tracklet tracks ans stubs to output channel
```
Typo: "tracklet tracks and stubs"

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/interface/ChannelAssignment.h:
```diff
@@ -0,0 +1,63 @@
+    ChannelAssignment(const edm::ParameterSet& iConfig, const tt::Setup* setup);
+    ~ChannelAssignment() {}
+    // sets channelId of given TTTrackRef, return false if track outside pt range
+    bool channelId(const TTTrackRef& ttTrackRef, int& channelId);
```
Since this class determines channels at both input & output of tracklet algo, the comment for channelId() should make clear that this is for "TrackBuilder or Duplicate Removal (if enabled).
Also, DR will always assign each track to a channel, so no inefficiency is caused to due finite Pt resolution. The lowest Pt bin will go down to Pt = 0.

tschuh replied:
first statement is done, second statement is false: f/w cannot represent infinite numbers.

tomalin commented on L1Trigger/TrackFindingTracklet/interface/ChannelAssignment.h:
```diff
@@ -0,0 +1,63 @@
+    // number of used channels
+    int numChannels() const { return numChannels_; }
+    // sets layerId of given TTStubRef and TTTrackRef, returns false if seeed stub
+    bool layerId(const TTTrackRef& ttTrackRef, const TTStubRef& ttStubRef, int& layerId);
```
Since layerId has many definitions, comment should make clear that this is "0-3 in the sequence of layers the seed type of this track projects to.".

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/interface/ChannelAssignment.h:
```diff
@@ -0,0 +1,63 @@
+    bool layerId(const TTTrackRef& ttTrackRef, const TTStubRef& ttStubRef, int& layerId);
+    // max number layers a sedd type may project to
+    int maxNumProjectionLayers() const { return maxNumProjectionLayers_; }
+    // map of used DTC tfp channels in InputRouter
```
Function channelEncoding() doesn't seem to be used anywhere?

tschuh replied:
if not in this PR then in the follow ups

tomalin commented on L1Trigger/TrackFindingTracklet/python/ChannelAssignment_cfi.py:
```diff
@@ -0,0 +1,34 @@
+ChannelAssignment_params = cms.PSet (
+
+  UseDuplicateRemoval = cms.bool   ( True ), # use tracklet seed type as channel id if False, binned track pt used if True
+  PtBoundaries        = cms.vdouble( 1.34 ), # pt Boundaries in GeV, last boundary is infinity
```
Why does PtBoundaries specify only 1 bin? The DR will use 12. Also comment should say these correspond to Pt bins used by DR.

tschuh replied:
these are actually 2, one for all positive and on one for all negative tracks. This defines the KF input channels. I assume that the DR merges their internal channel to this channel.

tschuh replied:
these are actually 2, one for all positive and on one for all negative tracks. This defines the KF input channels. I assume that the DR merges their internal channel to this channel.

tomalin replied:
ChannelAssignment_cfi.py should have comment at top making clear which channels it is configuring.

tomalin replied:
And wouldn't it make more sense to have 1/Pt boundaries than Pt boundaries? Then they can be (0, 1/1.34)

tschuh replied:
done and I prefer pt

tschuh committed:
name change of config parameter

tschuh committed:
code-format

tschuh committed:
namespace change and consitency check added.

tschuh commented:
I renamed the config parameter so to make clear that those are subsets of other parameter. I added a consistency check so that one cannot invent new seed types are projection layer within these subsets.

tomalin commented:
@tschuh Please check that you have responded to all comments, so this PR can be approved.

tschuh committed:
addressing ians comments

tschuh committed:
addressing ians comments

tomalin reviewed (approved):
Looks good.
